### PR TITLE
add get weights by parameter name for llama

### PR DESCRIPTION
**Pull request by zhaochenyang20**
## Motivation

To test the correctness of `update_parameter_from_distributed`, I add an test-only function to get the weights of a parameter in ModelRunner by the name of it.

## Checklist

- [X] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [X] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [X] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/server.py [hunk 1]
```diff
@@ -938,3 +970,59 @@ def stop_profile(self):
 
     async def get_server_info(self):
         return await _get_server_info()
+
+    def generate(
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: revert this

File: python/sglang/srt/utils.py [hunk 2]
```diff
@@ -922,6 +934,68 @@ def get_nvgpu_memory_capacity():
         )
 
 
+# Copy from pytorch and OpenRLHF to allow creating multiple main groups.
+# https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py
+def init_custom_process_group(
```
Comment (merrymercy) [inline python/sglang/srt/utils.py]: revert this

File: python/sglang/srt/server.py [hunk 3]
```diff
@@ -79,6 +80,7 @@
     assert_pkg_version,
     configure_logger,
     delete_directory,
+    init_custom_process_group,
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: revert this

File: test/srt/test_get_parameter_by_name.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    @classmethod
+    def setUpClass(cls):
+        cls.model = DEFAULT_SMALL_MODEL_NAME_FOR_TEST
+        cls.base_url = DEFAULT_URL_FOR_TEST.replace("2157", "1999")
+        cls.hf_model = AutoModelForCausalLM.from_pretrained(
+            cls.model, torch_dtype="bfloat16"
+        ).to("cuda:0")
+
+    @classmethod
+    def init_engine_and_server(cls, engine_tp, server_tp, engine_dp, server_dp):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]:
```suggestion
    def init_engine_and_server(cls, backend: str, dp: int, tp: int):
```
where backend can be `engine` or `server`

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: test/srt/test_get_parameter_by_name.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        if isinstance(ret, list) and len(ret) == 2:
+            print(f"running assert_allclose on data parallel")
+            np.testing.assert_allclose(ret[0], ret[1])
+            return np.array(ret[0])
+        return np.array(ret)
+
+    @classmethod
+    def test_update_weights_unexist_model(cls):
+        assert torch.cuda.device_count() >= 2, "At least 2 GPUs are required"
+        test_suits = [(1, 1, 1, 1), (2, 0, 1, 0), (0, 2, 0, 1)]
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: This test is not that important. To save CI time, probably just test one most easy case with one GPU.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/utils.py [hunk 6]
```diff
@@ -30,6 +30,7 @@
 import tempfile
 import time
 import warnings
+from datetime import timedelta
```
Comment (merrymercy) [inline python/sglang/srt/utils.py]:
```suggestion
```

File: python/sglang/srt/utils.py [hunk 7]
```diff
@@ -38,6 +39,7 @@
 import psutil
 import requests
 import torch
+import torch.distributed
```
Comment (merrymercy) [inline python/sglang/srt/utils.py]:
```suggestion
```

File: test/srt/test_get_parameter_by_name.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+    @staticmethod
+    def _process_return(ret):
+        if isinstance(ret, list) and len(ret) == 2:
+            print(f"running assert_allclose on data parallel")
+            np.testing.assert_allclose(ret[0], ret[1])
+            return np.array(ret[0])
+        return np.array(ret)
+
+    @classmethod
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: this should be not a classmethod

File: test/srt/test_get_parameter_by_name.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        if cls.process:
+            terminate_process(cls.process)
+
+    @classmethod
+    def tearDownClass(cls):
+        del cls.hf_model
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    @classmethod
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: make it a normal method

File: test/srt/test_get_parameter_by_name.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    def setUpClass(cls):
+        cls.model = DEFAULT_SMALL_MODEL_NAME_FOR_TEST
+        cls.base_url = DEFAULT_URL_FOR_TEST.replace("2157", "1999")
+        cls.hf_model = AutoModelForCausalLM.from_pretrained(
+            cls.model, torch_dtype="bfloat16"
+        ).to("cuda:0")
+
+    @classmethod
+    def init_backend(cls, backend, dp, tp):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: this does not need to be a classmethod

File: test/srt/test_get_parameter_by_name.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+                other_args=(
+                    "--tp-size",
+                    str(tp),
+                    "--dp-size",
+                    str(dp),
+                ),
+            )
+
+    @classmethod
+    def close_engine_and_server(cls):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: this does not need to be a classmethod

File: test/srt/test_get_parameter_by_name.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    def _process_return(ret):
+        if isinstance(ret, list) and len(ret) == 2:
+            print(f"running assert_allclose on data parallel")
+            np.testing.assert_allclose(ret[0], ret[1])
+            return np.array(ret[0])
+        return np.array(ret)
+
+    @classmethod
+    def test_update_weights_unexist_model(cls):
+        test_suits = [("Engine", 1, 1), ("Runtime", 1, 1)]
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py:96]: only run one 1-gpu case on CI.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/llama.py [hunk 13]
```diff
@@ -425,10 +427,76 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
...
+                    gathered_weights = [
+                        torch.zeros_like(weight) for _ in range(tp_size)
+                    ]
+                    torch.distributed.all_gather(gathered_weights, weight)
+                    weight = torch.cat(gathered_weights, dim=1)
+                return weight.cpu().to(torch.float32).numpy().tolist()[:truncate_size]
+            else:
+                return None
+
+        except Exception as e:
```
Comment (merrymercy) [inline python/sglang/srt/models/llama.py:500]: do not silently ignore the exception

File: test/srt/test_get_parameter_by_name.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    popen_launch_server,
+)
+from sglang.utils import terminate_process
+
+
+class TestUpdateWeights(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.model = DEFAULT_SMALL_MODEL_NAME_FOR_TEST
+        cls.base_url = DEFAULT_URL_FOR_TEST.replace("2157", "1999")
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: why do you need to replace this?

File: test/srt/test_get_parameter_by_name.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        cls.hf_model = AutoModelForCausalLM.from_pretrained(
+            cls.model, torch_dtype="bfloat16"
+        ).to("cuda:0")
+
+    def init_backend(cls, backend, dp, tp):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: the first argument should be `self`

File: test/srt/test_get_parameter_by_name.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+                timeout=DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
+                other_args=(
+                    "--tp-size",
+                    str(tp),
+                    "--dp-size",
+                    str(dp),
+                ),
+            )
+
+    def close_engine_and_server(cls):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: the first argument should be `self`

File: test/srt/test_get_parameter_by_name.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        if cls.engine:
+            cls.engine.shutdown()
+        if cls.process:
+            terminate_process(cls.process)
+
+    @classmethod
+    def tearDownClass(cls):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: move this just under `setUpClass`

File: test/srt/test_get_parameter_by_name.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        del cls.hf_model
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    def assert_update_weights_all_close(cls, param_name, truncate_size):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: self

File: test/srt/test_get_parameter_by_name.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+
+    @staticmethod
+    def _process_return(ret):
+        if isinstance(ret, list) and len(ret) == 2:
+            print(f"running assert_allclose on data parallel")
+            np.testing.assert_allclose(ret[0], ret[1])
+            return np.array(ret[0])
+        return np.array(ret)
+
+    def test_update_weights_unexist_model(cls):
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py]: self

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: test/srt/test_get_parameter_by_name.py (hunk 14 above)
Comment (zhaochenyang20) [reply test/srt/test_get_parameter_by_name.py]: Some time conflicts in CI. So I changed it.

**Review by merrymercy — APPROVED** [review]

File: test/srt/test_get_parameter_by_name.py [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+        ).to("cuda:0")
+
+    @classmethod
+    def tearDownClass(cls):
+        del cls.hf_model
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    def init_backend(self, backend, dp, tp):
+        self.engine = None
```
Comment (merrymercy) [inline test/srt/test_get_parameter_by_name.py:35]: Try to use less `self.` next time.

**Pull request merged**